src/main/scala/org/apache/spark/scheduler/cluster/armada/ArmadaClusterManagerBackend.scala

```diff
@@ -433,10 +430,13 @@ private[spark] class ArmadaClusterManagerBackend(
     }
   }
 
-  /** Get the count of pending executors.
+  /** Get the count of pending executors. Excludes executors that have already registered with
+    * Spark.
     */
   private[armada] def getPendingExecutorCount: Int = {
     pendingExecutors.synchronized {
+      val registeredIds = getExecutorIds().toSet
+      pendingExecutors --= registeredIds
       pendingExecutors.size
     }
```

Comment:
Before, executors were removed from `pendingExecutors` immediately when the pod started running in the `onExecutorRunning()`, even before they registered with Spark. 
This made the allocator’s gap calculation see pending = 0 and submit more executors. 
Moved the logic in getPendingExecutorCount(), which solves the issue of reducing the pendingExecutor counts before the executors get a chance to register themselves